File: Source/DriveWorks.Sdk.Examples.VbDotNet/MySpecificationMacroTask.vb
' Import the Specification namespace so we have access to Specification flow.
Imports DriveWorks.Specification

' Import EventFlow to access the Task EventFlow technology.
Imports DriveWorks.EventFlow

<Task("My Task", "embedded://DriveWorks.Sdk.Examples.VbDotNet.Puzzle-16x16.png", "SDK-Starter-Examples Plugin")>
Public Class MySpecificationMacroTask
    Inherits Task

    ' Register properties so DriveWorks can see them and build rules for them.
    Private ReadOnly mMyName As FlowProperty(Of String) = Me.Properties.RegisterStringProperty("Name", "Name of the constant to set.")
    Private ReadOnly mMyValue As FlowProperty(Of String) = Me.Properties.RegisterStringProperty("Value", "Value of the constant.")
    ' Register node output so DriveWorks can output data.
    Private ReadOnly myNodeOutput As NodeOutput = Me.Outputs.Register("Log", "My Task", GetType(String))

    Protected Overrides Sub Execute(ByVal ctx As SpecificationContext)
        Dim theConstant As ProjectConstant = Nothing

        ' Perform the main work of the task here.
        If ctx.Project.Constants.TryGetConstant(mMyName.Value, theConstant) Then

            ' Set the constant value. 
            theConstant.Value = mMyValue.Value
            ' Mark the Task as successful. 
            Me.SetState(NodeExecutionState.Successful, String.Format("The constant '{0}' value was successfully set.", mMyName.Value))
            'Set node output value
            myNodeOutput.Fulfill(String.Format("The constant '{0}' value was successfully set.", mMyName.Value))

        Else

            ' Mark the Task as failed and report that we could not find the constant.
            Me.SetState(NodeExecutionState.Failed, String.Format("The constant '{0}' does not exist.", mMyName.Value))
            'Set node output value
            myNodeOutput.Fulfill(String.Format("The constant '{0}' does not exist.", mMyName.Value))

        End If
    End Sub
End Class

File: Source/DriveWorks.Sdk.Examples.CSharp/MySpecificationMacroTask.cs
// Import the Specification namespace so we have access to Specification flow.
using DriveWorks.Specification;

// Import EventFlow to access the Task EventFlow technology. 
using DriveWorks.EventFlow;

namespace DriveWorks.Sdk.Examples.CSharp
{
    [Task("My Task", "embedded://DriveWorks.Sdk.Examples.CSharp.Puzzle-16x16.png", "SDK-Starter-Examples Plugin", true)]
    public class MySpecificationMacroTask : Task
    {
        private FlowProperty<string> MyName { get; }
        private FlowProperty<string> MyValue { get; }
        private NodeOutput MyNodeOutput { get; }

        public MySpecificationMacroTask()
        {
            // Register properties so DriveWorks can see them and build rules for them.
            MyName = this.Properties.RegisterStringProperty("Name", "Name of the constant to set.");
            MyValue = this.Properties.RegisterStringProperty("Value", "Value of the constant.");
            // Register node output so DriveWorks can output data.
            MyNodeOutput = this.Outputs.Register("Log", "My Task", typeof(string));
        }

        protected override void Execute(SpecificationContext ctx)
        {
            ProjectConstant theConstant = null;

            // Perform the main work of the task here.
            if (ctx.Project.Constants.TryGetConstant(MyName.Value, ref theConstant))
            {
                // Set the constant value.
                theConstant.Value = MyValue.Value;

                //Check if Value is empty or null
                if (string.IsNullOrEmpty(MyValue.Value))
                {
                    //Report Value is empty.
                    this.SetState(NodeExecutionState.SuccessfulWithWarnings, "The constant value was successfully cleared.");
                }
                else
                {
                    // Mark the Task as successful. 
                    this.SetState(NodeExecutionState.Successful, "The constant value was successfully set.");
                }

                //Set node output value
                MyNodeOutput.Fulfill(string.Format("The constant '{0}' value was successfully set.", MyName.Value));
            }
            else
            {
                // Mark the Task as failed and report that we could not find the constant.
                this.SetState(NodeExecutionState.Failed, string.Format("The constant '{0}' does not exist.", MyName.Value));

                // Set node output value
                MyNodeOutput.Fulfill(string.Format("The constant '{0}' does not exist.", MyName.Value));
            }
        }
    }
}